### PR TITLE
feat(vite): Add ability to switch Webpack for Vite on the web side

### PR DESCRIPTION
**Pull request by dac09**
### Experimental Vite Support 🎤

This PR does the following:

- [x] changes within the framework to make it Vite compatible (there's quite a lot here.... see full breakdown: https://www.notion.so/redwoodjs/Vite-RW-95f38a4e57ab4391b172b9d771b75104) 
- [x] adds setup command to opt-in to using Vite. This will modify your project to make it Vite compatible
- [x] updates CLI to look for the bundler flag, and switch commands when running dev
- [x] find a way to make storybook run - currently failing because of the `await import` in DevFatalErrorPage
		Added experiments.topLevelAwait to storybook webpack config! Storybook is still running webpack

**Importantly 👉** Everything should be backwards compatible (so no breaking changes for anyone remaining on webpack)

---

### Outstanding tasks

- [x] Use config hook and setup the preset as a Vite plugin
- [x] Can we "inject" the entry point automatically instead of modifying index.html
- [x] Redirect to root doesn't seem to be working (i.e. refresh the page after navigating, you get a blank page)
- [x] Swap build command to vite
- [x] Fix prerender failures caused by top level await in FatalErrorPage

### CODE REVIEW

File: packages/auth-providers-web/ambient.d.ts
Comment (Tobbe) [inline packages/auth-providers-web/ambient.d.ts:14]: This shouldn't be here. Created this PR to clean it up https://github.com/redwoodjs/redwood/pull/6968

File: packages/cli-helpers/src/lib/installHelpers.ts
Comment (Tobbe) [inline packages/cli-helpers/src/lib/installHelpers.ts:3]: Should this be part of this PR? I like the refactor, just doesn't seem to have anything to do with Vite

File: packages/auth-providers-web/ambient.d.ts
Comment (Tobbe) [inline packages/auth-providers-web/ambient.d.ts:5]:
```suggestion
```

File: packages/internal/src/build/babel/web.ts
Comment (Tobbe) [inline packages/internal/src/build/babel/web.ts:116]: Do we really support .jsx files?

File: packages/web/src/components/DevFatalErrorPage.tsx
Comment (Tobbe) [inline packages/web/src/components/DevFatalErrorPage.tsx:11]: This comment broke the connection between the RWJS_SRC_ROOT comment and the usage of RWJS_SRC_ROOT. Is it possible to move the @MARK somewhere else?

**Review by Tobbe — CHANGES_REQUESTED** [review]

File: packages/internal/src/build/babel/web.ts
Comment (dac09) [reply packages/internal/src/build/babel/web.ts:116]: Yes we do - read this for context: https://github.com/vitejs/vite/discussions/3448

I have a plugin temporarily so that you don't need to switch to .jsx, but once Vite is stable, I think we should move everything to .jsx as per the thread.

File: packages/cli-helpers/src/lib/installHelpers.ts
Comment (dac09) [reply packages/cli-helpers/src/lib/installHelpers.ts:3]: I need the addWebPackages function in the setup vite command.

File: packages/web/src/components/DevFatalErrorPage.tsx
Comment (dac09) [reply packages/web/src/components/DevFatalErrorPage.tsx:11]: Yeah I'll remove this, got it working.

File: tasks/framework-tools/lib/framework.mjs
Comment (dac09) [inline tasks/framework-tools/lib/framework.mjs:124]: @jtoar wanted to chat to you about this and the bin aliasing we do!

File: packages/auth-providers-web/ambient.d.ts
Comment (dac09) [reply packages/auth-providers-web/ambient.d.ts:14]: yeah, I'll take it out too! Do we need this ambient file in that case?

**Comment by dac09** [comment]
@KrisCoulson ready to test! How to try on your project before the PR is merged:


```
# Copy latest and greatest code from this branch
yarn rwfw project:sync 

# Setup vite in your project
yarn rw setup vite --no-add-package

# Inside redwood/packages/vite - in the framework.
# this will publish the new redwoodjs/vite package locally
yarn build && yarn dlx yalc publish

# Then inside your project
cd web && yarn dlx yalc add @redwoodjs/vite
yarn
```

File: packages/auth-providers-web/ambient.d.ts
Comment (Tobbe) [reply packages/auth-providers-web/ambient.d.ts:13]: I don't think so. But things also worked previously when we didn't have it, even though maybe we should have. Does Vite change something that makes the ambient files more important?

File: packages/cli-helpers/src/lib/installHelpers.ts
Comment (Tobbe) [reply packages/cli-helpers/src/lib/installHelpers.ts:3]: Cool. I always just thought of it as an auth specific helper. But of course other commands could use it too! Should we make an auth-cli-helpers package with things only used by the auth setup commands? I guess that kind of ties into the @redwoodjs/paths discussions we've been having. So you don't have to answer. Just something to keep in mind for when we do have the rw/paths conversation next time.

Also don't have to make any code changes here. As I said - I like the refactor 🙂 Just missed that you used it for the vite setup command.

File: packages/internal/src/build/babel/web.ts
Comment (Tobbe) [reply packages/internal/src/build/babel/web.ts:116]: I knew I remembered something about a js vs jsx conversation! https://github.com/redwoodjs/redwood/issues/734

Comment (Tobbe) [reply packages/internal/src/build/babel/web.ts:116]: So we do support .jsx, we just didn't want to make it the default 🙂  (That was a long time ago though, so people might have changed their name)

File: .eslintrc.js
Comment (jtoar) [inline .eslintrc.js:62]: This one's gone now:

```suggestion
```

File: packages/auth-providers/dbAuth/web/src/__tests__/dbAuth.test.ts
Comment (jtoar) [inline packages/auth-providers/dbAuth/web/src/__tests__/dbAuth.test.ts:8]: > // But because this is going to be moved as part of decoupling, ignoring for now

Not decoupling auth right? Some other decoupling? Just clarifying

File: packages/auth-providers/dbAuth/web/src/dbAuth.ts
Comment (jtoar) [inline packages/auth-providers/dbAuth/web/src/dbAuth.ts:56]: Why doesn't this one need `globalThis`? Not saying it should or shouldn't, I just don't understand the nuance

File: packages/web/src/components/DevFatalErrorPage.tsx
Comment (jtoar) [inline packages/web/src/components/DevFatalErrorPage.tsx:10]: I don't understand the nuance yet so just going to ask: why no `globalThis`?

File: packages/cli-helpers/src/lib/installHelpers.ts
Comment (jtoar) [inline packages/cli-helpers/src/lib/installHelpers.ts:17]: This one's changed a bit; the deleted code is up to date but not this one

```suggestion
export const addApiPackages = (apiPackages: string[]) => ({
  title: 'Adding required api packages...',
  task: async () => {
    await execa('yarn', ['workspace', 'api', 'add', ...apiPackages])
})
```

File: packages/cli/src/commands/generate/dbAuth/__tests__/dbAuth.test.js
Comment (jtoar) [inline packages/cli/src/commands/generate/dbAuth/__tests__/dbAuth.test.js:5]: Seems like we need the `fs` stuff back here: https://github.com/redwoodjs/redwood/actions/runs/3692167967/jobs/6250798520

File: packages/cli/src/commands/setup/vite/vite.js
Comment (jtoar) [inline packages/cli/src/commands/setup/vite/vite.js:40]: The handler and it's dependencies need to be in its own file, `viteHandler.js` (see `builderHandler.js`, `devHandler.js`, etc) and asynchronously imported here; else the upfront require cost is going to slow down other commands

File: packages/internal/src/build/babel/web.ts
Comment (jtoar) [reply packages/internal/src/build/babel/web.ts:116]: > ...once Vite is stable, I think we should move everything to .jsx as per the thread

We've discussed this once before and a lot of the team was against this. So queue it up for discussion again when you're ready

File: packages/vite/package.json
Comment (jtoar) [inline packages/vite/package.json:32]: `esbuild` is missing given it's used in the `src/index.ts` file (line 8):

```ts
import { transform } from 'esbuild'
```

File: packages/vite/src/index.ts
Comment (jtoar) [inline packages/vite/src/index.ts:17]: These functions are executed as soon as this file is imported, which may not be a problem given the changes I've made to the CLI to ensure the `cwd` is set before the CLI imports anything. But just something to note if you think this file could be executed somehow before `cwd` is set

**Review by jtoar — COMMENTED** [review]
Haven't tried it out yet and don't understand Vite at all so just commented on the things I could initially

File: packages/cli/src/commands/setup/vite/vite.js
Comment (jtoar) [reply packages/cli/src/commands/setup/vite/vite.js:40]: We haven't done this with many setup commands before, but I think we should continue the trend

File: packages/cli/src/commands/setup/vite/templates/vite.config.ts.template
Comment (jtoar) [inline packages/cli/src/commands/setup/vite/templates/vite.config.ts.template:11]: This leaves us with a bit of type duplication in the case of ts files; wonder if we could clean it up? I'll see if it's this way with other commands too

File: packages/cli/src/commands/setup/vite/vite.js
Comment (jtoar) [inline packages/cli/src/commands/setup/vite/vite.js:89]: Seems like this step ruins the formatting and deletes the comments:

```toml
[web]
title = "Redwood App"
port = 8_910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = [ ]
bundler = "vite"

[api]
port = 8_911

[browser]
open = true

```

File: packages/cli/src/commands/setup/vite/vite.js
Comment (jtoar) [inline packages/cli/src/commands/setup/vite/vite.js:101]: Not sure about the name here; why not just call it `index.js` like the one we already have? And this file has to be there? Right now we keep `index.js` under wraps, and give users the option to generate it if they want more power. What's the nuance or reasoning for changing?

**Review by jtoar — COMMENTED** [review]
Another quick round

File: packages/cli/src/commands/setup/vite/vite.js
Comment (jtoar) [reply packages/cli/src/commands/setup/vite/vite.js:40]: Got a PR here to get the other commands on the same page https://github.com/redwoodjs/redwood/pull/7156

File: packages/cli/src/commands/setup/vite/vite.js
Comment (dac09) [reply packages/cli/src/commands/setup/vite/vite.js:101]: I've intentionally called it entry-client, because for render modes you need a different entry point for server rendering. 

I know I originally hid the index.js, but in retrospect I think it doesn't change as often as I thought it did. We also didn't have codemod capability at the time!

File: packages/cli/src/commands/setup/vite/vite.js
Comment (dac09) [reply packages/cli/src/commands/setup/vite/vite.js:89]: Yeah noticed this too, but don't intend on solving right now.

File: packages/cli/src/commands/setup/vite/templates/vite.config.ts.template
Comment (dac09) [reply packages/cli/src/commands/setup/vite/templates/vite.config.ts.template:11]: ts-to-js doesn't remove these type comments, so its the only way to guarantee both. Do you see an error in your editor? Mine is fine with both the comment and type

File: packages/vite/src/index.ts
Comment (dac09) [reply packages/vite/src/index.ts:17]: Hmmm... I'll move this inside the function!

File: packages/web/src/components/DevFatalErrorPage.tsx
Comment (dac09) [reply packages/web/src/components/DevFatalErrorPage.tsx:10]: When you use `globalThis` - it fails to find the variable in either Webpack or Vite.

Directly accessing the variable like this is the only way I could make sure it works for both!

File: packages/auth-providers/dbAuth/web/src/__tests__/dbAuth.test.ts
Comment (dac09) [reply packages/auth-providers/dbAuth/web/src/__tests__/dbAuth.test.ts:8]: Hey I've fixed this... I misread the path, so thought it wouldn't be needed!

File: packages/vite/package.json
Comment (dac09) [reply packages/vite/package.json:32]: Esbuild is a dependency of vite, should be ok no?

File: packages/cli/src/commands/setup/vite/vite.js
Comment (dac09) [reply packages/cli/src/commands/setup/vite/vite.js:40]: Yeah happy to change this, no worries! Feel free to push if you like

File: packages/cli/src/commands/setup/vite/vite.js
Comment (dac09) [reply packages/cli/src/commands/setup/vite/vite.js:89]: Actually fixed it with a simple string replace!

**Comment by dac09** [comment]
@jtoar ready for merge me thinks - addressed all comments, and fixed the issues @KrisCoulson and I found the other day! Also bumped to Vite 4. Note that I see an issue in storybook that looks like an unrelated issue: https://github.com/storybookjs/storybook/pull/20383

It's still experimental ofcourse, and I will need to add fixes as we find them, but its now in a state where I think I'd like to get into more hands

FYI there's some stuff I've intentionally not done yet, e.g. 

- [ ] Vite dev and vite build should be called programatically
- [ ] Prerendering images causes images to reload 

Next step for me is to write up a guide for people to try over the holidays!

File: packages/auth-providers/dbAuth/web/src/dbAuth.ts
Comment (dac09) [reply packages/auth-providers/dbAuth/web/src/dbAuth.ts:52]: Thought I replied to this! Not adding the `globalThis` is the only way I can ensure it works with both Webpack and Vite.

I think there's subtle differences in the way these values are defined between the two bundlers

**Comment by dac09** [comment]
@Tobbe would you like to review again/dismiss the old review please? 

Want to get this into main (pushing up with merge conflicts resolved and new fixture shortly)

**Comment by Tobbe** [comment]
@dac09 I'll try to give it another look after our stand-up meeting later today

**Review by Tobbe — APPROVED** [review]
I'm happy with your changes/replies to my comments, but please let @jtoar go through his comments too before you merge

File: packages/cli/src/commands/setup/vite/templates/vite.config.ts.template
Comment (jtoar) [reply packages/cli/src/commands/setup/vite/templates/vite.config.ts.template:11]: Nah it's not about an error just about duplication; not something we need to clean up now

File: packages/vite/package.json
Comment (jtoar) [reply packages/vite/package.json]: It may be ok, but it may also be susceptible to hoisting since the internal package lists esbuild as a dependency. I'm just going off of yarn's rulebook which says "Packages should only ever require what they formally list in their dependencies" https://yarnpkg.com/advanced/rulebook#packages-should-only-ever-require-what-they-formally-list-in-their-dependencies

**Comment by jtoar** [comment]
I think we can see about listing esbuild as a dependency later. I'm inclined to just cause it's yarn's rules, but let's see what happens hoisting-wise once it's merged and the package is published. I'll also separate the handler after merge—I don't want to risk messing it up while it's in a workable state, and I'm only just getting back into things. One thing I'm wondering is if the toml change users have to make should just be a note logged to the console till the toml stuff is figured out. But that's also something we can do after merge.

Great stuff @dac09! I'll merge this so I can publish a canary version of the new package.

**Pull request merged**
